romancal/datamodels/library.py

```diff
@@ -0,0 +1,504 @@
+                    }
+                ],
+            }
+            self._members = self._asn["products"][0]["members"]
+
+            # _asn_dir?
+            # _asn_path?
+
+        elif isinstance(init, self.__class__):
+            # TODO clone/copy?
+            raise NotImplementedError()
+        else:
+            raise NotImplementedError()
+
+        # make sure first model is loaded in memory (as expected by stpipe)
+        if self._asn_n_members == 1:
+            # FIXME stpipe also reaches into _models (instead of _model_store)
+            self._models = [self._load_member(0)]
+
+    def __del__(self):
+        # FIXME when stpipe no longer uses '_models'
+        if hasattr(self, "_models"):
+            self._models[0].close()
+
+    @property
+    def asn(self):
+        # return a "read only" association
+        def _to_read_only(obj):
+            if isinstance(obj, dict):
+                return MappingProxyType(obj)
+            if isinstance(obj, list):
+                return tuple(obj)
+            return obj
+
+        return asdf.treeutil.walk_and_modify(self._asn, _to_read_only)
+
+    # TODO we may want to not expose this as it could go out-of-sync
+    # pretty easily with the actual models.
+    # @property
+    # def members(self):
+    #     return self.asn['products'][0]['members']
+
+    @property
+    def group_names(self):
+        names = set()
+        for member in self._members:
+            names.add(member["group_id"])
+        return names
+
+    @property
+    def group_indices(self):
+        group_dict = {}
+        for i, member in enumerate(self._members):
+            group_id = member["group_id"]
+            if group_id not in group_dict:
+                group_dict[group_id] = []
+            group_dict[group_id].append(i)
+        return group_dict
+
+    def __len__(self):
+        return len(self._members)
+
+    def __getitem__(self, index):
+        if not self._open:
+            raise ClosedLibraryError("ModelLibrary is not open")
+
+        # if model was already borrowed, raise
+        if index in self._ledger:
+            raise BorrowError("Attempt to double-borrow model")
+
+        if index in self._model_store:
+            model = self._model_store[index]
+        else:
+            model = self._load_member(index)
+            if not self._on_disk:
+                # it's ok to keep this in memory since _on_disk is False
+                self._model_store[index] = model
+
+        # track the model is "in use"
+        self._ledger[index] = model
+        return model
+
+    def __setitem__(self, index, model):
+        if index not in self._ledger:
+            raise BorrowError("Attempt to return non-borrowed model")
+
+        # un-track this model
+        del self._ledger[index]
+
+        # and store it
+        self._model_store[index] = model
+
+        # TODO should we allow this to change group_id for the member?
+
+    def discard(self, index, model):
```

Comment:
It seems to me it would be nice to be able to do both of the following:
```
with library:
    library.discard(0)
    library[0].discard()
```
Are both of these supported?

Also, I would expect discarding a model to remove it from the library entirely, including the temporary file where it is stored.  Does that happen?

Reply:
Good point. I had similar thoughts about the `discard` naming. In this case `discard` means to return a model to the library but not write it out to a (possibly new) temporary file. The model will still be in the library and can be later "borrowed" again). The main use is when you only need to read something from a model (like the wcs) but don't want to incur the cost of writing out the model when it's returned to the library. Do you think it would be clearer if this modified vs un-modified "return" was an argument to `return_model` (or `shelve`)? Perhaps something like:
```python
with library:
    model = library.borrow(0)  # borrow the 0th model
    # do some read-only stuff with the model
    library.shelve(0, model, modified=False)   # with modified=True being the default
```
To address the other part of your comment supporting `library.discard(0)` sounds like a good idea and perhaps `library.discard(model)`. I can't currently think of way of supporting `library[0].discard` without patching a `discard` method onto the model. I briefly considered this but I worry the added complexity would be hard to get ride and prone to bugs.

Reply:
Is there a straightforward way to check if the model has been modified when it is shelved, and only incur that cost if so?  My worry would be that a random developer or user would be unaware of the modified keyword (e.g., after having copied a random example from elsewhere in the code base) and just leave it as default, even though they do not intend to modify the model.

Reply:
Also, I see your point about `library[0].discard` and I think your proposed `library.discard(model)` would be a suitable replacement

Reply:
I'm not aware of a straightforward way to check if a model has been modified. The risk you mention about copying an example is a good point. For catching that type of error I think we can rely on the code reviews and tests. Depending on how we expose "on_disk" (what it will default to and if it will be dynamic) we may want to unit test every step that uses the library for both `on_disk=True` and `on_disk=False` (or specifically disallow one mode). This might be an argument for making `on_disk=False` the default and only setting it by a pipeline/step parameter (and making sure that the PR that enables it tests the code in both modes).